File: Projetos/ACBrLib/Demos/C#/BAL/Imports/Dinamico/MT/ACBrBAL.cs
using System.Text;
using ACBrLib.Core;

namespace ACBrLib.BAL
{
    public sealed partial class ACBrBAL : ACBrLibHandle
    {
        #region Constructors

        public ACBrBAL(string eArqConfig = "", string eChaveCrypt = "") : base(IsWindows ? "ACBrBAL64.dll" : "libacbrbal64.so",
                                                                               IsWindows ? "ACBrBAL32.dll" : "libacbrbal32.so")
        {
            var inicializar = GetMethod<BAL_Inicializar>();
            var ret = ExecuteMethod(() => inicializar(ref libHandle, ToUTF8(eArqConfig), ToUTF8(eChaveCrypt)));

            CheckResult(ret);

            Config = new BALConfig(this);
        }

        #endregion Constructors

        #region Properties

        public string Nome
        {
            get
            {
                var bufferLen = BUFFER_LEN;
                var buffer = new StringBuilder(bufferLen);

                var method = GetMethod<BAL_Nome>();
                var ret = ExecuteMethod(() => method(libHandle, buffer, ref bufferLen));

                CheckResult(ret);

                return ProcessResult(buffer, bufferLen);
            }
        }

        public string Versao
        {
            get
            {
                var bufferLen = BUFFER_LEN;
                var buffer = new StringBuilder(bufferLen);

                var method = GetMethod<BAL_Versao>();
                var ret = ExecuteMethod(() => method(libHandle, buffer, ref bufferLen));

                CheckResult(ret);

                return ProcessResult(buffer, bufferLen);
            }
        }

        public BALConfig Config { get; }

        #endregion Properties

        #region Methods

        #region Ini

        public override void ConfigGravar(string eArqConfig = "")
        {
            var gravarIni = GetMethod<BAL_ConfigGravar>();
            var ret = ExecuteMethod(() => gravarIni(libHandle, ToUTF8(eArqConfig)));

            CheckResult(ret);
        }

        public override void ConfigLer(string eArqConfig = "")
        {
            var lerIni = GetMethod<BAL_ConfigLer>();
            var ret = ExecuteMethod(() => lerIni(libHandle, ToUTF8(eArqConfig)));

            CheckResult(ret);
        }

        public override T ConfigLerValor<T>(ACBrSessao eSessao, string eChave)
        {
            var method = GetMethod<BAL_ConfigLerValor>();

            var bufferLen = BUFFER_LEN;
            var pValue = new StringBuilder(bufferLen);
            var ret = ExecuteMethod(() => method(libHandle, ToUTF8(eSessao.ToString()), ToUTF8(eChave), pValue, ref bufferLen));
            CheckResult(ret);

            var value = ProcessResult(pValue, bufferLen);
            return ConvertValue<T>(value);
        }

        public override void ConfigGravarValor(ACBrSessao eSessao, string eChave, object value)
        {
            if (value == null) return;

            var method = GetMethod<BAL_ConfigGravarValor>();
            var propValue = ConvertValue(value);

            var ret = ExecuteMethod(() => method(libHandle, ToUTF8(eSessao.ToString()), ToUTF8(eChave), ToUTF8(propValue)));
            CheckResult(ret);
        }

        public override void ImportarConfig(string eArqConfig = "")
        {
            var importarConfig = GetMethod<BAL_ConfigImportar>();
            var ret = ExecuteMethod(() => importarConfig(libHandle, ToUTF8(eArqConfig)));

            CheckResult(ret);
        }

        public override string ExportarConfig()
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<BAL_ConfigExportar>();
            var ret = ExecuteMethod(() => method(libHandle, buffer, ref bufferLen));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen);
        }

        #endregion Ini

        public void Ativar()
        {
            var method = GetMethod<BAL_Ativar>();
            var ret = ExecuteMethod(() => method(libHandle));

            CheckResult(ret);
        }

        public void Desativar()
        {
            var method = GetMethod<BAL_Desativar>();
            var ret = ExecuteMethod(() => method(libHandle));

            CheckResult(ret);
        }

        public decimal LePeso(int MillisecTimeOut = 1000)
        {
            var peso = 0D;
            var method = GetMethod<BAL_LePeso>();
            var ret = ExecuteMethod(() => method(libHandle, MillisecTimeOut, ref peso));

            CheckResult(ret);

            return (decimal)peso;
        }

        public void SolicitarPeso()
        {
            var method = GetMethod<BAL_SolicitarPeso>();
            var ret = ExecuteMethod(() => method(libHandle));

            CheckResult(ret);
        }

        public decimal UltimoPesoLido()
        {
            var peso = 0D;
            var method = GetMethod<BAL_UltimoPesoLido>();
            var ret = ExecuteMethod(() => method(libHandle, ref peso));

            CheckResult(ret);

            return (decimal)peso;
        }

        public decimal InterpretarRespostaPeso(string resposta)
        {
            var peso = 0D;
            var method = GetMethod<BAL_InterpretarRespostaPeso>();
            var ret = ExecuteMethod(() => method(libHandle, ToUTF8(resposta), ref peso));

            CheckResult(ret);

            return (decimal)peso;
        }

        #region Private Methods

        protected override void FinalizeLib()
        {
            var finalizar = GetMethod<BAL_Finalizar>();
            var ret = ExecuteMethod(() => finalizar(libHandle));
            CheckResult(ret);
        }

        protected override string GetUltimoRetorno(int iniBufferLen = 0)
        {
            var bufferLen = iniBufferLen < 1 ? BUFFER_LEN : iniBufferLen;
            var buffer = new StringBuilder(bufferLen);
            var ultimoRetorno = GetMethod<BAL_UltimoRetorno>();

            if (iniBufferLen < 1)
            {
                ExecuteMethod(() => ultimoRetorno(libHandle, buffer, ref bufferLen));
                if (bufferLen <= BUFFER_LEN) return FromUTF8(buffer);

                buffer.Capacity = bufferLen;
            }

            ExecuteMethod(() => ultimoRetorno(libHandle, buffer, ref bufferLen));
            return FromUTF8(buffer);
        }

        #endregion Private Methods

        #endregion Methods
    }
}

File: Projetos/ACBrLib/Demos/Java/BAL/Imports/ACBrLibBALMT/src/com/acbr/bal/ACBrBAL.java
package com.acbr.bal;

import com.acbr.ACBrLibBase;
import com.acbr.ACBrSessao;
import com.sun.jna.Library;
import com.sun.jna.Native;
import com.sun.jna.Platform;
import com.sun.jna.Pointer;
import com.sun.jna.ptr.DoubleByReference;
import com.sun.jna.ptr.FloatByReference;
import com.sun.jna.ptr.IntByReference;
import com.sun.jna.ptr.PointerByReference;
import java.io.File;
import java.nio.ByteBuffer;
import java.nio.file.Paths;

public final class ACBrBAL extends ACBrLibBase implements AutoCloseable {
    
    private interface ACBrBALLib extends Library {
        
        static String JNA_LIBRARY_NAME = LibraryLoader.getLibraryName();
        public final static ACBrBALLib INSTANCE = LibraryLoader.getInstance();
        
        class LibraryLoader {
            
            private static String library = "";
            private static ACBrBALLib instance = null;
                        
            private static String getLibraryName() {
                if ( library.isEmpty() ) {
                    if(Platform.isWindows()){
                        library = Platform.is64Bit() ? "ACBrBAL64" : "ACBrBAL32";                        
                    }else{
                        library = Platform.is64Bit() ? "acbrbal64" : "acbrbal32";
                    }
                    
                }  
                return library;
            }
            
            public static ACBrBALLib getInstance() {
                if ( instance == null ) {
                    instance = (ACBrBALLib) Native.synchronizedLibrary(
                            (Library) Native.loadLibrary(JNA_LIBRARY_NAME, ACBrBALLib.class));
                }
                
                return instance;
            }                             
        }
    
        int BAL_Inicializar(PointerByReference libHandler, String eArqConfig, String eChaveCrypt);

        int BAL_Finalizar(Pointer libHandler);

        int BAL_Nome(Pointer libHandler, ByteBuffer buffer, IntByReference bufferSize);

        int BAL_Versao(Pointer libHandler, ByteBuffer buffer, IntByReference bufferSize);

        int BAL_UltimoRetorno(Pointer libHandler, ByteBuffer buffer, IntByReference bufferSize);

        int BAL_ConfigLer(Pointer libHandler, String eArqConfig);

        int BAL_ConfigGravar(Pointer libHandler, String eArqConfig);

        int BAL_ConfigLerValor(Pointer libHandler, String eSessao, String eChave, ByteBuffer buffer, IntByReference bufferSize);

        int BAL_ConfigGravarValor(Pointer libHandler, String eSessao, String eChave, String valor);

        int BAL_Ativar(Pointer libHandler);

        int BAL_Desativar(Pointer libHandler);

        int BAL_LePeso(Pointer libHandler, int MillisecTimeOut, DoubleByReference Peso);

        int BAL_SolicitarPeso(Pointer libHandler);

        int BAL_UltimoPesoLido(Pointer libHandler, DoubleByReference Peso);

        int BAL_ConfigImportar(Pointer libHandler, String eArqConfig);
        
		int BAL_ConfigExportar(Pointer libHandler, ByteBuffer buffer, IntByReference bufferSize);
        
        int BAL_InterpretarRespostaPeso(Pointer libHandler, ByteBuffer eResposta, FloatByReference Peso);                   
    }

    public ACBrBAL() throws Exception{
        File iniFile = Paths.get(System.getProperty("user.dir"), "ACBrLib.ini").toFile();
        if ( !iniFile.exists() ) {
            iniFile.createNewFile();
        }
        
        PointerByReference handle = new PointerByReference();
        int ret = ACBrBALLib.INSTANCE.BAL_Inicializar(handle, toUTF8(iniFile.getAbsolutePath()), toUTF8(""));
        checkResult(ret);
		setHandle(handle.getValue());
    }
    
    public ACBrBAL(String eArqConfig, String eChaveCrypt) throws Exception{
        PointerByReference handle = new PointerByReference();
        int ret = ACBrBALLib.INSTANCE.BAL_Inicializar(handle, toUTF8(eArqConfig), toUTF8(eChaveCrypt));
        checkResult(ret);
		setHandle(handle.getValue());
    }
    
    @Override
    public void close() throws Exception {
        int ret = ACBrBALLib.INSTANCE.BAL_Finalizar(getHandle());
        checkResult(ret);
    }
    
    @Override
    protected void finalize() throws Throwable {
        try {
            int ret = ACBrBALLib.INSTANCE.BAL_Finalizar(getHandle());
            checkResult(ret);
        } catch (Exception e) {
            super.finalize();
        }
    }
    
    public String nome() throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrBALLib.INSTANCE.BAL_Nome(getHandle(), buffer, bufferLen);
        checkResult(ret);

        return fromUTF8(buffer, bufferLen.getValue());
    }

    public String versao() throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrBALLib.INSTANCE.BAL_Versao(getHandle(), buffer, bufferLen);
        checkResult(ret);

        return fromUTF8(buffer, bufferLen.getValue());
    }

    public void configLer() throws Exception {
        configLer("");
    }

    public void configLer(String eArqConfig) throws Exception {
        int ret = ACBrBALLib.INSTANCE.BAL_ConfigLer(getHandle(), toUTF8(eArqConfig));
        checkResult(ret);
    }

    public void configGravar() throws Exception {
        configGravar("");
    }

    public void configGravar(String eArqConfig) throws Exception {
        int ret = ACBrBALLib.INSTANCE.BAL_ConfigGravar(getHandle(), toUTF8(eArqConfig));
        checkResult(ret);
    }

    @Override
    public String configLerValor(ACBrSessao eSessao, String eChave) throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrBALLib.INSTANCE.BAL_ConfigLerValor(getHandle(), toUTF8(eSessao.name()), toUTF8(eChave), buffer, bufferLen);
        checkResult(ret);
        
        return processResult(buffer, bufferLen);
    }

    @Override
    public void configGravarValor(ACBrSessao eSessao, String eChave, Object value) throws Exception {
        int ret = ACBrBALLib.INSTANCE.BAL_ConfigGravarValor(getHandle(), toUTF8(eSessao.name()), toUTF8(eChave), toUTF8(value.toString()));
        checkResult(ret);
    }
    
    public void ativar() throws Exception {
        int ret = ACBrBALLib.INSTANCE.BAL_Ativar(getHandle());
        checkResult(ret);
    }
    
    public void desativar() throws Exception {
        int ret = ACBrBALLib.INSTANCE.BAL_Desativar(getHandle());
        checkResult(ret);
    }
    
    public double lePeso(int MillisecTimeOut) throws Exception {
        DoubleByReference peso = new DoubleByReference(STR_BUFFER_LEN);

        int ret = ACBrBALLib.INSTANCE.BAL_LePeso(getHandle(), MillisecTimeOut, peso);
        checkResult(ret);       
        
        System.out.println(peso.getValue());
        
        return peso.getValue();
    }

    public void SolicitarPeso() throws Exception {
        int ret = ACBrBALLib.INSTANCE.BAL_SolicitarPeso(getHandle());
        checkResult(ret);
    }

    public double ultimoPesoLido() throws Exception {
        DoubleByReference peso = new DoubleByReference(STR_BUFFER_LEN);

        int ret = ACBrBALLib.INSTANCE.BAL_UltimoPesoLido(getHandle(), peso);
        checkResult(ret);
        
        return peso.getValue();
    }

    public float interpretarRespostaPeso(ByteBuffer eResposta, double var) throws Exception {
        FloatByReference peso = new FloatByReference(STR_BUFFER_LEN);

        int ret = ACBrBALLib.INSTANCE.BAL_InterpretarRespostaPeso(getHandle(), eResposta, peso);
        checkResult(ret);
        
        return peso.getValue();
    } 
    
    public void ConfigImportar(String eArqConfig) throws Exception {
        
        int ret = ACBrBALLib.INSTANCE.BAL_ConfigImportar(getHandle(), eArqConfig);
        checkResult(ret);
        
    }
    
    public String ConfigExportar() throws Exception {
		
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrBALLib.INSTANCE.BAL_ConfigExportar(getHandle(), buffer, bufferLen);
        checkResult(ret);

        return fromUTF8(buffer, bufferLen.getValue());
		
    }
    
    @Override
    protected void UltimoRetorno(ByteBuffer buffer, IntByReference bufferLen) {
        ACBrBALLib.INSTANCE.BAL_UltimoRetorno(getHandle(), buffer, bufferLen);
    }       
}
